File: First-Edition/BookSamples/CH05/src/Cellular/Assets/Automata/Seeds.hs
{-# LANGUAGE DataKinds #-}
{-# LANGUAGE PolyKinds #-}
{-# LANGUAGE FlexibleInstances #-}
{-# LANGUAGE TypeApplications #-}
module Cellular.Assets.Automata.Seeds where

import Cellular.Language.Board
import Cellular.Language.Algorithm
import Cellular.Language.Automaton
import Cellular.Assets.Automata.LifeLike


-- Seeds (B2/S)
type Neighbors2 = 'NeighborsCount A '[2]

type SeedsStep = 'Step @LifeLikeStates ('DefState D)
  '[ 'StateTransition D A Neighbors2  -- "Born rule"
   ]

type SeedsRule = 'Rule
  "Seeds"
  "seeds"
  ('AdjacentsLvl 1)
  SeedsStep

File: First-Edition/BookSamples/rosetta-stone/scala3/rsc2p3_type_level_interfaces/src/main/scala/cellular/assets/Seeds.scala
package cellular
package assets

import graninas.typelevel._
import cellular.language._
import cellular.extensions._


object Seeds {

  type A = State["Alive", 1]
  type D = State["Dead", 0]

  type Neighbors2 = NeighborsCount[D, 2 :: IN]

  type Trans1 = StateTransition[D, A, Neighbors2]

  type Transitions =
    Cons[IStateTransition, Trans1,
      Nil[IStateTransition]]


  type SeedsStep = Step[D, Transitions]

  type States = Cons[IState, A, Cons[IState, D, Nil[IState]]]

  type SeedsRule = Rule[
    "Seeds",
    "seeds",
    AdjacentsLvl[1],
    SeedsStep]

}
